src/Documentation/Summarizer/summarize_documentation.py

```diff
@@ -53,6 +54,76 @@ def items_of_kind(items, kind):
         )
     ]
 
+def summaries_table(namespace, kind, header):
+    table_header = f"""
+## {header}
+
+| Name | Summary |
+|------|---------|
+"""
+    def first_line(summary):
+        lines = summary.splitlines()
+        return lines[0] if lines else ""
+
+    return (
+        table_header + "\n".join(
+            f"|[{item['name']}](xref:{item['uid']}) |{first_line(item['summary'])} |"
+            for item in namespace[kind]
+        )
+        if kind in namespace and namespace[kind]
+        else ""
+    )
+
+def write_namespace_summary(target_file : Path, namespace):
+    """
+    Given the contents of a namespace, either creates a new Markdown file
+    summarizing those contents, or appends those contents to an existing
+    Markdown file for that namespace.
+    """
+
+    begin_region = "<!-- summaries -->"
+    end_region = "<!-- /summaries -->"
+
+    # Make the summary table to be injected.
+    summaries = begin_region + "\n" + "\n".join([
+        summaries_table(namespace, 'operations', 'Operations'),
+        summaries_table(namespace, 'functions', 'Functions'),
+        summaries_table(namespace, 'newtypes', 'User-defined types'),
+    ]) + "\n" + end_region + "\n"
+
+
+    # Check if the file already exists
+    if target_file.exists():
+        contents = target_file.read_text(encoding='utf8')
+        # If it does exist, look for <!-- summaries --> and <!-- /summaries -->
+        # comments, and if they exist, replace inside that section.
+        if begin_region in contents:
+            contents = contents[:contents.index(begin_region)] + summaries + contents[contents.index(end_region) + len(end_region):]
+        else:
+            contents += "\n" + summaries
+
+        with open(target_file, 'w', encoding='utf8') as f:
+            f.write(contents)
+
+    # If the file does not exist, go on and make it now.
+    else:
+        with open(target_file, 'w', encoding='utf8') as f:
```

Comment:
There's a time-of-check to time-of-use bug here. What if the file is created after `target_file.exists()` but before opening it for writing? Maybe it doesn't matter that much for this script though.

Reply:
Agreed, and I did consider that but it made the flow of the code a bit less obvious to avoid that issue. Given that this script is intended to run on a static set of assets (e.g.: build outputs), I think that those assets changing during summarization is likely out of scope for this tool.